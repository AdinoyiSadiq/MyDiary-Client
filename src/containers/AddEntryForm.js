import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createEntry } from '../actions/entryActions';
import validateAuth from '../lib/validation';

const initialState = {
  title: '',
  content: '',
  touched: {
    title: false,
    content: false,
  },
};

const fieldNames = ['title', 'content'];

class AddEntryForm extends Component {
  state = initialState;

  handleSubmit = async (event) => {
    event.preventDefault();
    const { createEntry, history } = this.props;
    const { title, content } = this.state;

    const validationError = validateAuth({ title, content }, fieldNames);
    const changedTouchState = this.changeTouchState(fieldNames, true);
    this.setState({ touched: changedTouchState });

    if (!validationError.status) {
      createEntry({ title, content }, (entryId) => history.push(`/main/entry/${entryId}`));
      this.setState(initialState);
    }
  }

  handleChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  }
  
  handleBlur = (field) => () => {
    const { touched } = this.state;
    this.setState({
      touched: { ...touched, [field]: true },
    });
  }

  changeTouchState = (fields, state) => {
    const touchState = {};
    fields.forEach((field) => {
      touchState[field] = state;
    });
    return touchState;
  }

  cancelEntry = () => {
    const { history } = this.props;
    this.setState({initialState});
    history.push('/main/entries')
  }

  render() {
    const { title, content, touched } = this.state;
    const { errorMessage } = this.props;
    const error = validateAuth({ title, content }, fieldNames);
    
    return (
      <main>
        <section>
          <form onSubmit={this.handleSubmit}>
          
            <div className="heading">Add an Entry</div>
            <div className="entryError">{errorMessage}</div>
            <input 
              type="text" 
              placeholder="Title" 
              id="title" 
              value={title} 
              onChange={this.handleChange}
              onBlur={this.handleBlur('title')}
            />
            <div className="entryError">{touched.title ? error.title : ''}</div>
            <textarea 
              rows="20" 
              cols="50" 
              placeholder="Content" 
              id="content" 
              value={content} 
              onChange={this.handleChange}
              onBlur={this.handleBlur('content')}
            />
            <div className="entryError">{touched.content ? error.content : ''}</div>
            <button type="submit">Save</button>
            <button type="button" onClick={this.cancelEntry}>Cancel</button>
          </form>
        </section>
      </main>
    )
  }
}

function mapStateToProps(state) {
  return { errorMessage: state.entry.errorMessage };
}

export default connect(mapStateToProps, { createEntry })(AddEntryForm);