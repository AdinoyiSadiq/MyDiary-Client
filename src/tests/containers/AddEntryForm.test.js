import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import AddEntryForm from '../../containers/AddEntryForm';
import Root from '../../root';

let wrapped;
let wrapped1;

const type = 'update';
const match = { 
  params: { id: 2 } 
}

beforeEach(() => {
  wrapped = mount(
    <Root>
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <AddEntryForm />
      </MemoryRouter>
    </Root>,
  );

  wrapped1 = mount(
    <Root>
      <MemoryRouter initialEntries={[{ key: 'testKey' }]}>
        <AddEntryForm type={type} match={match} />
      </MemoryRouter>
    </Root>,
  );
});

afterEach(() => wrapped.unmount());
afterEach(() => wrapped1.unmount());

describe('Add Entry Form UI', () => {
  describe('render features', () => {
    test('container should render as expected', () => {
      const tree = toJson(wrapped);
      expect(tree).toMatchSnapshot();
    });

    test('container should render as expected when in update mode', () => {
      const tree = toJson(wrapped1);
      expect(tree).toMatchSnapshot();
    });

    it('has two input fields and a button', () => {
      expect(wrapped.find('input').length).toEqual(1);
      expect(wrapped.find('textarea').length).toEqual(1);
      expect(wrapped.find('button').length).toEqual(2);
    });
  });

  describe('when typing into fields', () => {
    beforeEach(() => {
      wrapped.find('input').first().simulate('change', { target: { id: 'title', value: 'title' } });
      wrapped.find('textarea').first().simulate('change', { target: { id: 'content', value: 'content' } });
      wrapped.update();
    });

    it('shows that text has been entered into the title field', () => {
      expect(wrapped.find('input').first().prop('value')).toEqual('title');
    });

    it('shows that text has been entered into the content field', () => {
      expect(wrapped.find('textarea').first().prop('value')).toEqual('content');
    });

    it('empties fields when form is submitted', () => {
      wrapped.find('form').simulate('submit');
      wrapped.update();

      expect(wrapped.find('input').first().prop('value')).toEqual('');
      expect(wrapped.find('textarea').first().prop('value')).toEqual('');
    });
  });
});

describe('Add Entry Form client-side validation', () => {
  beforeEach(() => {
    wrapped.find('input').first().simulate('change', { target: { id: 'title', value: '' } });
    wrapped.find('textarea').first().simulate('change', { target: { id: 'content', value: '' } });
    wrapped.update();
  });

  it('shows error messages if both fields are empty', () => {
    wrapped.find('form').simulate('submit');
    wrapped.update();

    expect(wrapped.find('.entryError').at(1).text()).toEqual('Please enter a title');
    expect(wrapped.find('.entryError').at(2).text()).toEqual('Please enter some content');
  });
});