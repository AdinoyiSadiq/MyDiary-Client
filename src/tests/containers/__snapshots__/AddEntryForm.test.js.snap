// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Entry Form UI render features container should render as expected 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(AddEntryForm)>
          <AddEntryForm
            clearEntryError={[Function]}
            createEntry={[Function]}
            errorMessage=""
            getSingleEntry={[Function]}
            updateEntry={[Function]}
          >
            <main>
              <section>
                <form
                  onSubmit={[Function]}
                >
                  <div
                    className="heading"
                  >
                    Add an Entry
                  </div>
                  <div
                    className="entryError"
                  />
                  <input
                    id="title"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
                  />
                  <div
                    className="entryError"
                  />
                  <textarea
                    cols="50"
                    id="content"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Content"
                    rows="20"
                    value=""
                  />
                  <div
                    className="entryError"
                  />
                  <button
                    type="submit"
                  >
                    Save
                  </button>
                  <button
                    onClick={[Function]}
                    type="button"
                  >
                    Cancel
                  </button>
                </form>
              </section>
            </main>
          </AddEntryForm>
        </Connect(AddEntryForm)>
      </Router>
    </MemoryRouter>
  </Provider>
</_default>
`;

exports[`Add Entry Form UI render features container should render as expected when in update mode 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testKey",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testKey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testKey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Connect(AddEntryForm)
          match={
            Object {
              "params": Object {
                "id": 2,
              },
            }
          }
          type="update"
        >
          <AddEntryForm
            clearEntryError={[Function]}
            createEntry={[Function]}
            errorMessage=""
            getSingleEntry={[Function]}
            match={
              Object {
                "params": Object {
                  "id": 2,
                },
              }
            }
            type="update"
            updateEntry={[Function]}
          >
            <main>
              <section>
                <form
                  onSubmit={[Function]}
                >
                  <div
                    className="heading"
                  >
                    Update Entry
                  </div>
                  <div
                    className="entryError"
                  />
                  <input
                    id="title"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Title"
                    type="text"
                    value=""
                  />
                  <div
                    className="entryError"
                  />
                  <textarea
                    cols="50"
                    id="content"
                    onBlur={[Function]}
                    onChange={[Function]}
                    placeholder="Content"
                    rows="20"
                    value=""
                  />
                  <div
                    className="entryError"
                  />
                  <button
                    type="submit"
                  >
                    Save
                  </button>
                  <button
                    onClick={[Function]}
                    type="button"
                  >
                    Cancel
                  </button>
                </form>
              </section>
            </main>
          </AddEntryForm>
        </Connect(AddEntryForm)>
      </Router>
    </MemoryRouter>
  </Provider>
</_default>
`;
